### PR TITLE
Fortran to C++: build buffer masks

### PR DESCRIPTION
**Pull request by EZoni**
Goal of this PR is to implement the Fortran subroutine `warpx_build_buffer_masks` in C++:

- implementation tested in 2D and 3D using input file similar to the one for test `subcyclingMR` (with non-zero buffer sizes);
- results obtained with Fortran and C++ implementations compared with the AMReX tool `fcompare`;
- old Fortran implementation removed from the repository.

### CODE REVIEW

File: Source/WarpX.cpp [hunk 1]
```diff
@@ -1280,17 +1280,57 @@ WarpX::BuildBufferMasks ()
...
+{
+    bool setnull;
+    const auto lo = amrex::lbound( tbx );
+    const auto hi = amrex::ubound( tbx );
+#if (AMREX_SPACEDIM == 2)
+    int k = lo.z;
+    for     (int j = lo.y; j <= hi.y; ++j) {
+        for (int i = lo.x; i <= hi.x; ++i) {
+            setnull = false;
+            for     (int jj = j-ng; jj <= j+ng; ++jj) {
```
Comment (MaxThevenet) [inline Source/WarpX.cpp]:
```suggestion
            // If gmsk=0 for any neighbor with ng cells, current cell is in the buffer region.
            for     (int jj = j-ng; jj <= j+ng; ++jj) {
```

File: Source/WarpX.cpp [hunk 2]
```diff
@@ -1280,17 +1280,57 @@ WarpX::BuildBufferMasks ()
...
+            if ( setnull ) msk(i,j,k) = 0;
+            else           msk(i,j,k) = 1;
+        }
+    }
+#elif (AMREX_SPACEDIM == 3)
+    for         (int k = lo.z; k <= hi.z; ++k) {
+        for     (int j = lo.y; j <= hi.y; ++j) {
+            for (int i = lo.x; i <= hi.x; ++i) {
+                setnull = false;
+                for         (int kk = k-ng; kk <= k+ng; ++kk) {
```
Comment (MaxThevenet) [inline Source/WarpX.cpp]:
```suggestion
                // If gmsk=0 for any neighbor with ng cells, current cell is in the buffer region.
                for         (int kk = k-ng; kk <= k+ng; ++kk) {
```

File: Source/WarpX.H [hunk 3]
```diff
@@ -560,6 +560,8 @@ private:
     void LoadBalance ();
 
     void BuildBufferMasks ();
+    void BuildBufferMasksInBox ( const amrex::Box tbx, amrex::Array4<int> &msk,
```
Comment (MaxThevenet) [inline Source/WarpX.H]: Could you add a Doxygen-readable docstring to this function, either here or in the cpp file?

File: Source/WarpX.cpp [hunk 4]
```diff
@@ -1280,17 +1280,57 @@ WarpX::BuildBufferMasks ()
...
-                    const Box& tbx = mfi.growntilebox();
-                    warpx_build_buffer_masks (BL_TO_FORTRAN_BOX(tbx),
-                                              BL_TO_FORTRAN_ANYD((*bmasks)[mfi]),
-                                              BL_TO_FORTRAN_ANYD(tmp[mfi]),
-                                              &ngbuffer);
+                    const Box tbx = mfi.growntilebox();
+                    Array4<int>  msk = (*bmasks)[mfi].array();
+                    Array4<int> gmsk = tmp[mfi].array();
+                    // Build buffer mask within current FArrayBox
+                    BuildBufferMasksInBox( tbx, msk, gmsk, ngbuffer );
```
[the remark this answers is not in the document]
Comment (EZoni) [reply Source/WarpX.cpp]: Yes, this is a good suggestion and I tried this before choosing the current implementation, but the compiler complained about the fact that I try to use a "deleted function". As I understand it, the reason is that the usage of the member function is disabled by appending the `delete` specifier to the end of the function declaration in `AMReX_IArrayBox.H`: `IArrayBox (const IArrayBox&) = delete;` (line 76). Still need to figure out how to fix this if we choose the implementation that you suggest. Would you have any ideas on this?

[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/WarpX.cpp]:
```c++
void
WarpX::BuildBufferMasks ()
{
[...]
                for (MFIter mfi(*bmasks, true); mfi.isValid(); ++mfi)
		{
                    const Box tbx = mfi.growntilebox();
                    // Build buffer mask within current FArrayBox                                                                                                                                                                                                                                                             
                    // msk gmsk                                                                                                                                                                                                                                                                                               
                    BuildBufferMasksInBox( tbx, (*bmasks)[mfi], tmp[mfi], ngbuffer );
                }
[...]
}

void
WarpX::BuildBufferMasksInBox ( const Box tbx, IArrayBox &buffer_mask, const IArrayBox& guard_mask, const int ng )
{
    Array4<int>  msk = buffer_mask.array();
    Array4<int const> gmsk = guard_mask.array();
[...]
}
```
works fine for me.

[the remark this answers is not in the document]
Comment (EZoni) [reply Source/WarpX.cpp]: Thanks, sorry, lack of C++ knowledge on my side: was passing guard_mask not by reference. Should be implemented correctly now.

File: Source/WarpX.cpp [hunk 5]
```diff
@@ -1280,17 +1279,67 @@ WarpX::BuildBufferMasks ()
...
+    const auto lo = amrex::lbound( tbx );
+    const auto hi = amrex::ubound( tbx );
+    Array4<int> msk = buffer_mask.array();
+    Array4<int const> gmsk = guard_mask.array();
+#if (AMREX_SPACEDIM == 2)
+    int k = lo.z;
+    for     (int j = lo.y; j <= hi.y; ++j) {
+        for (int i = lo.x; i <= hi.x; ++i) {
+            setnull = false;
+            // If gmsk=0 for any neighbor with ng cells, current cell is in the buffer region
```
Comment (MaxThevenet) [inline Source/WarpX.cpp]:
```suggestion
            // If gmsk=0 for any neighbor within ng cells, current cell is in the buffer region
```
sorry about that.

File: Source/WarpX.cpp [hunk 6]
```diff
@@ -1280,17 +1279,67 @@ WarpX::BuildBufferMasks ()
...
+            if ( setnull ) msk(i,j,k) = 0;
+            else           msk(i,j,k) = 1;
+        }
+    }
+#elif (AMREX_SPACEDIM == 3)
+    for         (int k = lo.z; k <= hi.z; ++k) {
+        for     (int j = lo.y; j <= hi.y; ++j) {
+            for (int i = lo.x; i <= hi.x; ++i) {
+                setnull = false;
+                // If gmsk=0 for any neighbor with ng cells, current cell is in the buffer region
```
Comment (MaxThevenet) [inline Source/WarpX.cpp]:
```suggestion
                // If gmsk=0 for any neighbor within ng cells, current cell is in the buffer region
```

**Comment by EZoni** [comment]
> Wow that many files where `WarpX_f.H` was included but useless, even for `DO_ELECTROSTATIC` and `USE_PSATD_PICSAR`? That's pretty cool ^^.

Yes, but only for `DO_ELECTROSTATIC=TRUE`, while `USE_PSATD_PICSAR=TRUE` seems not to be supported anymore, as we discussed briefly on Slack.

**Pull request merged**